### ISSUE TITLE
swintransformer模型export后使用deploy下的predict_cls.py验证不准确

### ISSUE THREAD

**Issue by BerryRB** [PaddlePaddle/PaddleClas#1633]
1. PaddleClas版本以及PaddlePaddle版本：PaddleClas release/2.3和PaddlePaddle 2.2.1
 2. 训练环境信息：
  a. 具体操作系统，如Linux
  b. Python版本号，如Python3.7
  c. CUDA/cuDNN版本， 如CUDA10.1
 3. 模型：SwinTransformer_base_patch4_window7_224

问题：训练了SwinTransformer_base_patch4_window7_224模型，训练模型保存后，使用tools文件夹下的infer.py脚本进行预测，预测类别及得分如下：
![pcls_issue1](https://user-images.githubusercontent.com/34699510/148877311-7b25889d-da6c-4ba0-9b2e-fe7fe3619e57.png)

使用以下命令进行模型导出：
`python tools/export_model.py -c ./deploy/models/ppswin_b224_0106/SwinTransformer_base_patch4_window7_224.yaml -o weights=output/SwinTransformer_base_patch4_window7_224/best_model`

导出后获得inference模型，使用以下命令对同一张图片进行预测：
`python python/predict_cls.py \
    -c configs/inference_cls.yaml \
    -o Global.infer_imgs=/home/dbsb-data/appdbsb9/clstrain/odnega/nega_00088.jpg \
    -o Global.inference_model_dir=../inference/ \
    -o PostProcess.Topk.class_id_map_file=None`
预测类别及得分如下：
![pcls_issue2](https://user-images.githubusercontent.com/34699510/148877541-7494ac43-2919-4e30-9aaa-1ee205f09a6c.png)

麻烦看下什么原因，同样的环境的代码文件，导出的ResNet50模型则不存在此问题。

**Comment by Intsigstephon** [comment]
好的，感谢反馈这个问题；训练swintransforer的配置文件可以提供下吗？

**Comment by BerryRB** [comment]
> 好的，感谢反馈这个问题；训练swintransforer的配置文件可以提供下吗？

`# global configs
Global:
  checkpoints: null
  pretrained_model: null
  output_dir: ./output/
  device: gpu
  save_interval: 1
  eval_during_train: True
  eval_interval: 1
  epochs: 40
  print_batch_step: 50
  use_visualdl: True
  # used for static mode and model export
  image_shape: [ 3, 224, 224 ]
  save_inference_dir: ./inference
  # training model under @to_static
  to_static: False

# model architecture
Arch:
  name: SwinTransformer_base_patch4_window7_224
  class_num: 2

# loss function config for traing/eval process
Loss:
  Train:
    - CELoss:
        weight: 1.0
        epsilon: 0.1
  Eval:
    - CELoss:
        weight: 1.0


Optimizer:
  name: AdamW
  beta1: 0.9
  beta2: 0.999
  epsilon: 1e-8
  weight_decay: 0.05
  no_weight_decay_name: absolute_pos_embed relative_position_bias_table .bias norm
  one_dim_param_no_weight_decay: True
  lr:
    name: Cosine
    learning_rate: 1e-4
    eta_min: 5e-6
    warmup_epoch: 5
    warmup_start_lr: 1e-6


# data loader for train and eval
DataLoader:
  Train:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/train_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - RandCropImage:
            size: 224
            interpolation: bicubic
            backend: pil
        - RandFlipImage:
            flip_code: 1
        - TimmAutoAugment:
            config_str: rand-m9-mstd0.5-inc1
            interpolation: bicubic
            img_size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
        - RandomErasing:
            EPSILON: 0.25
            sl: 0.02
            sh: 1.0/3.0
            r1: 0.3
            attempt: 10
            use_log_aspect: True
            mode: pixel
      batch_transform_ops:
        - OpSampler:
            MixupOperator:
              alpha: 0.8
              prob: 0.5
            CutmixOperator:
              alpha: 1.0
              prob: 0.5

    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: True
    loader:
      num_workers: 4
      use_shared_memory: True

  Eval:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/val_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - ResizeImage:
            resize_short: 256
            interpolation: bicubic
            backend: pil
        - CropImage:
            size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: False
    loader:
      num_workers: 4
      use_shared_memory: True

Infer:
  infer_imgs: /home/dbsb-data/appdbsb9/clstrain/odnega/nega_00006.jpg
  batch_size: 8
  transforms:
    - DecodeImage:
        to_rgb: True
        channel_first: False
    - ResizeImage:
        resize_short: 256
        interpolation: bicubic
        backend: pil
    - CropImage:
        size: 224
    - NormalizeImage:
        scale: 1.0/255.0
        mean: [ 0.485, 0.456, 0.406 ]
        std: [ 0.229, 0.224, 0.225 ]
        order: ''
    - ToCHWImage:
  PostProcess:
    name: Topk
    topk: 1
    class_id_map_file: /home/dbsb-data/appdbsb9/clstrain/label_list.txt

Metric:
  Eval:
    - TopkAcc:
        topk: [ 1 ]
`

你好，这是配置文件

**Comment by BerryRB** [comment]
**
# global configs
Global:
  checkpoints: null
  pretrained_model: null
  output_dir: ./output/
  device: gpu
  save_interval: 1
  eval_during_train: True
  eval_interval: 1
  epochs: 40
  print_batch_step: 50
  use_visualdl: True
  # used for static mode and model export
  image_shape: [ 3, 224, 224 ]
  save_inference_dir: ./inference
  # training model under @to_static
  to_static: False

# model architecture
Arch:
  name: SwinTransformer_base_patch4_window7_224
  class_num: 2

# loss function config for traing/eval process
Loss:
  Train:
    - CELoss:
        weight: 1.0
        epsilon: 0.1
  Eval:
    - CELoss:
        weight: 1.0


Optimizer:
  name: AdamW
  beta1: 0.9
  beta2: 0.999
  epsilon: 1e-8
  weight_decay: 0.05
  no_weight_decay_name: absolute_pos_embed relative_position_bias_table .bias norm
  one_dim_param_no_weight_decay: True
  lr:
    name: Cosine
    learning_rate: 1e-4
    eta_min: 5e-6
    warmup_epoch: 5
    warmup_start_lr: 1e-6


# data loader for train and eval
DataLoader:
  Train:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/train_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - RandCropImage:
            size: 224
            interpolation: bicubic
            backend: pil
        - RandFlipImage:
            flip_code: 1
        - TimmAutoAugment:
            config_str: rand-m9-mstd0.5-inc1
            interpolation: bicubic
            img_size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
        - RandomErasing:
            EPSILON: 0.25
            sl: 0.02
            sh: 1.0/3.0
            r1: 0.3
            attempt: 10
            use_log_aspect: True
            mode: pixel
      batch_transform_ops:
        - OpSampler:
            MixupOperator:
              alpha: 0.8
              prob: 0.5
            CutmixOperator:
              alpha: 1.0
              prob: 0.5

    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: True
    loader:
      num_workers: 4
      use_shared_memory: True

  Eval:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/val_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - ResizeImage:
            resize_short: 256
            interpolation: bicubic
            backend: pil
        - CropImage:
            size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: False
    loader:
      num_workers: 4
      use_shared_memory: True

Infer:
  infer_imgs: /home/dbsb-data/appdbsb9/clstrain/odnega/nega_00006.jpg
  batch_size: 8
  transforms:
    - DecodeImage:
        to_rgb: True
        channel_first: False
    - ResizeImage:
        resize_short: 256
        interpolation: bicubic
        backend: pil
    - CropImage:
        size: 224
    - NormalizeImage:
        scale: 1.0/255.0
        mean: [ 0.485, 0.456, 0.406 ]
        std: [ 0.229, 0.224, 0.225 ]
        order: ''
    - ToCHWImage:
  PostProcess:
    name: Topk
    topk: 1
    class_id_map_file: /home/dbsb-data/appdbsb9/clstrain/label_list.txt

Metric:
  Eval:
    - TopkAcc:
        topk: [ 1 ]

**

**Comment by BerryRB** [comment]
> 好的，感谢反馈这个问题；训练swintransforer的配置文件可以提供下吗？

```text
# global configs
Global:
  checkpoints: null
  pretrained_model: null
  output_dir: ./output/
  device: gpu
  save_interval: 1
  eval_during_train: True
  eval_interval: 1
  epochs: 40
  print_batch_step: 50
  use_visualdl: True
  # used for static mode and model export
  image_shape: [ 3, 224, 224 ]
  save_inference_dir: ./inference
  # training model under @to_static
  to_static: False

# model architecture
Arch:
  name: SwinTransformer_base_patch4_window7_224
  class_num: 2

# loss function config for traing/eval process
Loss:
  Train:
    - CELoss:
        weight: 1.0
        epsilon: 0.1
  Eval:
    - CELoss:
        weight: 1.0


Optimizer:
  name: AdamW
  beta1: 0.9
  beta2: 0.999
  epsilon: 1e-8
  weight_decay: 0.05
  no_weight_decay_name: absolute_pos_embed relative_position_bias_table .bias norm
  one_dim_param_no_weight_decay: True
  lr:
    name: Cosine
    learning_rate: 1e-4
    eta_min: 5e-6
    warmup_epoch: 5
    warmup_start_lr: 1e-6


# data loader for train and eval
DataLoader:
  Train:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/train_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - RandCropImage:
            size: 224
            interpolation: bicubic
            backend: pil
        - RandFlipImage:
            flip_code: 1
        - TimmAutoAugment:
            config_str: rand-m9-mstd0.5-inc1
            interpolation: bicubic
            img_size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
        - RandomErasing:
            EPSILON: 0.25
            sl: 0.02
            sh: 1.0/3.0
            r1: 0.3
            attempt: 10
            use_log_aspect: True
            mode: pixel
      batch_transform_ops:
        - OpSampler:
            MixupOperator:
              alpha: 0.8
              prob: 0.5
            CutmixOperator:
              alpha: 1.0
              prob: 0.5

    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: True
    loader:
      num_workers: 4
      use_shared_memory: True

  Eval:
    dataset:
      name: ImageNetDataset
      image_root: /home/dbsb-data/appdbsb9/clstrain/
      cls_label_path: /home/dbsb-data/appdbsb9/clstrain/val_list.txt
      transform_ops:
        - DecodeImage:
            to_rgb: True
            channel_first: False
        - ResizeImage:
            resize_short: 256
            interpolation: bicubic
            backend: pil
        - CropImage:
            size: 224
        - NormalizeImage:
            scale: 1.0/255.0
            mean: [ 0.485, 0.456, 0.406 ]
            std: [ 0.229, 0.224, 0.225 ]
            order: ''
    sampler:
      name: DistributedBatchSampler
      batch_size: 8
      drop_last: False
      shuffle: False
    loader:
      num_workers: 4
      use_shared_memory: True

Infer:
  infer_imgs: /home/dbsb-data/appdbsb9/clstrain/odnega/nega_00006.jpg
  batch_size: 8
  transforms:
    - DecodeImage:
        to_rgb: True
        channel_first: False
    - ResizeImage:
        resize_short: 256
        interpolation: bicubic
        backend: pil
    - CropImage:
        size: 224
    - NormalizeImage:
        scale: 1.0/255.0
        mean: [ 0.485, 0.456, 0.406 ]
        std: [ 0.229, 0.224, 0.225 ]
        order: ''
    - ToCHWImage:
  PostProcess:
    name: Topk
    topk: 1
    class_id_map_file: /home/dbsb-data/appdbsb9/clstrain/label_list.txt

Metric:
  Eval:
    - TopkAcc:
        topk: [ 1 ]
```

你好，这是训练时的配置文件

**Comment by Intsigstephon** [comment]
有试过官网提供的预训练模型，infer结果和export_model之后的推理结果是否一致吗？可以先用pretrain模型试下哈

**Comment by Intsigstephon** [comment]
导出后获得inference模型，使用以下命令对同一张图片进行预测：
python python/predict_cls.py \ -c configs/inference_cls.yaml \ -o Global.infer_imgs=/home/dbsb-data/appdbsb9/clstrain/odnega/nega_00088.jpg \ -o Global.inference_model_dir=../inference/ \ -o PostProcess.Topk.class_id_map_file=None
预测类别及得分如下：
pcls_issue2

另外，这部分描述的命令和实际使用的命令不一致；麻烦再确认下所使用的inference模型路径是否正确

**Comment by BerryRB** [comment]
> python/predict_cls.py \ -c configs/inference_cls.yaml \ -o Global.infer_imgs=/home/dbsb-data/appdbsb9/clstrain/odnega/nega_00088.jpg \ -o Global.inference_model_dir=../inference/ \ -o PostProcess.Topk.class_id_map_file=None
> 预测类别及得分如下：
> pcls_issue2

使用官网提供的预训练模型结果是一致的：
![image](https://user-images.githubusercontent.com/34699510/148882396-a0c51197-4b70-4618-8aa2-b497831eac5b.png)

![image](https://user-images.githubusercontent.com/34699510/148882417-029947d8-fadb-4bb1-9056-4fba9140d5bf.png)

**Comment by BerryRB** [comment]
> 导出后获得inference模型，使用以下命令对同一张图片进行预测： python python/predict_cls.py \ -c configs/inference_cls.yaml \ -o Global.infer_imgs=/home/dbsb-data/appdbsb9/clstrain/odnega/nega_00088.jpg \ -o Global.inference_model_dir=../inference/ \ -o PostProcess.Topk.class_id_map_file=None 预测类别及得分如下： pcls_issue2
> 
> 另外，这部分描述的命令和实际使用的命令不一致；麻烦再确认下所使用的inference模型路径是否正确

这个是描述贴错了，实际命令是截图中的，模型路径确认过了，是正确的

**Comment by Intsigstephon** [comment]
需要你的模型来做复现，方便留个联系方式吗

**Comment by BerryRB** [comment]
> 需要你的模型来做复现，方便留个联系方式吗

好的感谢，18801297970，微信手机同号

**Comment by weiluonika** [comment]
请问问题解决了吗 
我是用paddleclas2.3 模型EfficientNetB4  使用自定义的训练集  导出模型与训练模型结果也不一致
